server/controllers/reports_proposed/data/variation_exploitation.js

```diff
@@ -0,0 +1,180 @@
+// reports_proposed/data/bilan.js
+// Collects and aggregates data for the enterprise bilan
+
+var q       = require('q');
+var db      = require('../../../lib/db');
+var numeral = require('numeral');
+
+var formatDollar = '$0,0.00', variationExploitationDate = new Date(), accountList = [],
+    subAccounts = [], finalList = [], previousExploitations = [],
+    currentExploitations = [], previousSubAccounts = [],
+    INCOME_EXPENSE_ACCOUNT_ID = 1, TITLE_ACCOUNT_ID = 3, BALANCE_ACCOUNT_ID = 2 ;
+
+// expose the http route
+exports.compile = function (options) {
+  'use strict';
+  var i18VariationExploitation = options.language == 'fr' ? require('../lang/fr.json').VARIATION_EXPLOITATION : require('../lang/en.json').VARIATION_EXPLOITATION;
+  var deferred = q.defer(), context = {}, infos = {}, ROOT_ACCOUNT_ID = 0;
+
+  context.i18n = i18VariationExploitation;
+  context.options = options;
+
+  var sql = 
+    'SELECT account.id, account.account_number, account.account_txt, account.parent, account.is_charge, account.account_type_id, (CASE WHEN account.is_charge = 1 THEN SUM(period_total.debit - period_total.credit) ELSE SUM(period_total.credit - period_total.debit) END) AS balance ' +
+    'FROM account LEFT JOIN period_total on account.id = period_total.account_id WHERE account.is_ohada = 1 AND account.account_type_id in (?,?) AND (period_total.fiscal_year_id in (?) OR ISNULL(period_total.fiscal_year_id)) ' +
+    'AND account.id NOT IN (SELECT account.id FROM account WHERE account.account_type_id = ?) GROUP BY account.id';
+
+  db.exec(sql, [INCOME_EXPENSE_ACCOUNT_ID, TITLE_ACCOUNT_ID, options.fy, BALANCE_ACCOUNT_ID])
+  .then(function (ce){ 
+    currentExploitations = ce || [];
+    return db.exec(sql, [INCOME_EXPENSE_ACCOUNT_ID, TITLE_ACCOUNT_ID, options.pfy, BALANCE_ACCOUNT_ID]);
+  })
+  .then(function (pe){
+    previousExploitations = pe || [];
+    context.data = process(currentExploitations, previousExploitations).sort(function (a, b){ return a.account_number - b.account_number;});
+    deferred.resolve(context);
```

Comment:
Oh, I found an excellent resource that describes this issue.  It's called the [Deferred Anti-Pattern](https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns#the-deferred-anti-pattern).  An "anti-pattern" is a way that people write code that is either not recommended or wrong.